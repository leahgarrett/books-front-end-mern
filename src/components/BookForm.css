.book-form {
    margin-left: 20px;
}
form {
    font-size: 16px;
    max-width: 500px;

}
input {
    margin: 5px 0 25px 0;
    display: block;
    width: 100%;
    font-size: 16px;
  }
  
label {
    display: block;
    width: 100%;
  }
  
  button {
    width: 120px;
    padding: 10px 30px;
  }