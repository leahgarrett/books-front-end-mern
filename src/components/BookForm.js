import React, { Component } from 'react';
import './BookForm.css';

class BookForm extends Component {
    constructor(props) {
        super(props);
        this.state = { 
            title: props.title || '',
            author: props.author || '',
            genre: props.genre || '',
            price: props.price || 0,
            id: props.id || 0
        };
      }

  checkValidPrice = () => {
    // require two decimal places
    var regex  = /^\d+(?:\.\d{0,2})$/;
    if (regex.test(this.state.price)){
      return true;
    }
    console.log('Invalid number - requires two characters after decimal')
    return false;
  }

  checkExists = () => {
    if ( this.state.title.length > 0 
      && this.state.author.length > 0  
      && this.state.genre.length > 0 
      && (this.state.price.length > 0)
    ) {
      return true;
    }
    console.log('Missing field')
    return false;
  }

  checkValid = () => {
    return this.checkExists() && this.checkValidPrice() 
  }

  handleClick = (e) => {
    e.preventDefault();

    if (this.checkValid()) {
      const newBook = {
          title: this.state.title, 
          author: this.state.author, 
          genre: this.state.genre, 
          price: Number(this.state.price), // convert to number before saving in DB
          id: this.state.id 
      }
      this.props.addNewBook(newBook)
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.id]: e.target.value, errorMessages: [] })
  }

  render(){
      return (
        <div className="book-form">
          <h1>Book Form</h1>
          <form>
            <label htmlFor="title">Title</label>
              <input 
                onChange={this.handleChange} 
                type="text" 
                id="title" 
                placeholder="title" 
                value={this.state.title} 
              />
              <label htmlFor="author">Author</label>
              <input 
                onChange={this.handleChange} 
                type="text" 
                id="author" 
                placeholder="author" 
                value={this.state.author} 
              />
              <label htmlFor="genre">Genre</label>
              <input 
                onChange={this.handleChange} 
                type="text" 
                id="genre" 
                placeholder="genre" 
                value={this.state.genre} 
              />
              <label htmlFor="price">Price</label>
              <input 
                onChange={this.handleChange} 
                type="number" 
                id="price" 
                placeholder="price" 
                value={this.state.price} 
              />

              <button onClick={this.handleClick}>Save</button>
              <button onClick={this.handleCancel}>Cancel</button>
          </form>
        </div>
      );
    }
}

export default BookForm;

