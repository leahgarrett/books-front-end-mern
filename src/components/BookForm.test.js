import React from "react";
import Enzyme, { shallow, render, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { createSerializer } from "enzyme-to-json";
import sinon from "sinon";

import BookForm from "./BookForm";



describe('<BookForm />', () => {
const books3 = [{
  author: "Ms. Remington Fahey",
  genre: "comedy",
  id: 1,
  price: 53.7,
  title: "Legacy Intelligent",
  _id: "5d19cd74a0c3e2c7f68ff800"
},
{
  author: "Isaac Steuber",
  genre: "thriller",
  id: 2,
  price: 20.99,
  title: "Incredible Incredible Plastic Computer",
  _id: "5d19cd74a0c3e2c7f68ff801"
},
{
  author: "Kennedy Howell",
  genre: "thriller",
  id: 3,
  price: 68.61,
  title: "compressing Soft mint green"
}
];
    it("renders component when there is no book data", () => {
        const wrapper = render(
            <BookForm 
                addNewBook={() => {}} 
             />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it("renders component when there is valid book data", () => {
        const wrapper = render(
            <BookForm 
                addNewBook={() => {}} 
                title={books3[0].title} 
                author={books3[0].author} 
                genre={books3[0].genre} 
                price={books3[0].price} 
             />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it("renders component when there is partial book data", () => {
        const wrapper = render(
            <BookForm 
                addNewBook={() => {}} 
                title={books3[0].title} 
                author={books3[0].author} 
             />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it("Does call add book when there is valid book data", () => {
        const spy = sinon.spy();
        const wrapper = mount(
            <BookForm 
                addNewBook={() => {spy()}} 
                title={books3[0].title} 
                author={books3[0].author} 
                genre={books3[0].genre} 
                price={books3[0].price} 
             />
        );
            
        wrapper.find("button").first().simulate("click");
        expect(spy.calledOnce).toBe(true);
    });

    it("Does not call add book when there is a missing title", () => {
        const spy = sinon.spy();
        const wrapper = mount(
            <BookForm 
                addNewBook={() => {spy()}} 
                author={books3[0].author} 
                genre={books3[0].genre} 
                price={books3[0].price} 
             />
        );
            
        wrapper.find("button").first().simulate("click");
        expect(spy.calledOnce).toBe(false);
    });

    it("Does not call add book when there is a missing author", () => {
        // add test
    });

    it("Does not call add book when there is a missing genre", () => {
        // add test
    });

    it("Does not call add book when there is a missing price", () => {
        // add test
    });

    it("Does not call add book when there is an invalid price", () => {
        // add test eg: 10e
    });

    it("Does not call add book when there is a negative price", () => {
        // add test 
    });
});
  