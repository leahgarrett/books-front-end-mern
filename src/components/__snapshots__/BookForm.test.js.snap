// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookForm /> renders component when there is no book data 1`] = `
<div
  class="book-form"
>
  <h1>
    Book Form
  </h1>
  <form>
    <label
      for="title"
    >
      Title
    </label>
    <input
      id="title"
      placeholder="title"
      type="text"
      value=""
    />
    <label
      for="author"
    >
      Author
    </label>
    <input
      id="author"
      placeholder="author"
      type="text"
      value=""
    />
    <label
      for="genre"
    >
      Genre
    </label>
    <input
      id="genre"
      placeholder="genre"
      type="text"
      value=""
    />
    <label
      for="price"
    >
      Price
    </label>
    <input
      id="price"
      placeholder="price"
      type="number"
      value="0"
    />
    <button>
      Save
    </button>
    <button>
      Cancel
    </button>
  </form>
</div>
`;

exports[`<BookForm /> renders component when there is partial book data 1`] = `
<div
  class="book-form"
>
  <h1>
    Book Form
  </h1>
  <form>
    <label
      for="title"
    >
      Title
    </label>
    <input
      id="title"
      placeholder="title"
      type="text"
      value="Legacy Intelligent"
    />
    <label
      for="author"
    >
      Author
    </label>
    <input
      id="author"
      placeholder="author"
      type="text"
      value="Ms. Remington Fahey"
    />
    <label
      for="genre"
    >
      Genre
    </label>
    <input
      id="genre"
      placeholder="genre"
      type="text"
      value=""
    />
    <label
      for="price"
    >
      Price
    </label>
    <input
      id="price"
      placeholder="price"
      type="number"
      value="0"
    />
    <button>
      Save
    </button>
    <button>
      Cancel
    </button>
  </form>
</div>
`;

exports[`<BookForm /> renders component when there is valid book data 1`] = `
<div
  class="book-form"
>
  <h1>
    Book Form
  </h1>
  <form>
    <label
      for="title"
    >
      Title
    </label>
    <input
      id="title"
      placeholder="title"
      type="text"
      value="Legacy Intelligent"
    />
    <label
      for="author"
    >
      Author
    </label>
    <input
      id="author"
      placeholder="author"
      type="text"
      value="Ms. Remington Fahey"
    />
    <label
      for="genre"
    >
      Genre
    </label>
    <input
      id="genre"
      placeholder="genre"
      type="text"
      value="comedy"
    />
    <label
      for="price"
    >
      Price
    </label>
    <input
      id="price"
      placeholder="price"
      type="number"
      value="53.7"
    />
    <button>
      Save
    </button>
    <button>
      Cancel
    </button>
  </form>
</div>
`;
